File: examples/foundry-local-python/provider.py
"""Adapt a Foundry Local native chat client to Vifu Runtime."""

from __future__ import annotations

from typing import Any, Iterable, Protocol

from vifu import AgentResponse, VifuRuntime


class FoundryChatClient(Protocol):
    def complete_streaming_chat(self, messages: list[dict[str, str]]) -> Iterable[Any]: ...


def register_foundry_agent(
    runtime: VifuRuntime,
    client: FoundryChatClient,
    *,
    model: str,
    endpoint: str = "foundry-chat",
) -> VifuRuntime:
    def invoke(request):
        prompt = request.input["prompt"]
        messages = [{"role": "user", "content": prompt}]
        chunks = iter(client.complete_streaming_chat(messages))
        parts: list[str] = []

        with request.trace.stage("first_token", metadata={"model": model}):
            for chunk in chunks:
                content = _chunk_content(chunk)
                if content:
                    parts.append(content)
                    request.trace.output_delta({"text": content})
                    break

        with request.trace.stage("decode", metadata={"model": model}):
            for chunk in chunks:
                content = _chunk_content(chunk)
                if content:
                    parts.append(content)
                    request.trace.output_delta({"text": content})

        return AgentResponse(
            output={"text": "".join(parts)},
            metadata={"model": model, "provider": "foundry-local"},
        )

    return runtime.agent(
        "foundry-local",
        invoke,
        endpoint=endpoint,
        metadata={"model": model, "framework": "foundry-local"},
    )


def _chunk_content(chunk: Any) -> str:
    choices = getattr(chunk, "choices", None)
    if not choices:
        return ""
    return getattr(choices[0].delta, "content", None) or ""

File: examples/foundry-local-typescript/provider.ts
import type {
  VifuAgentRequest,
  VifuAgentTrace,
  VifuRuntime,
} from "../../npm-packages/sdk/dist/index.js";

interface FoundryChunk {
  choices?: Array<{ delta?: { content?: string | null } }>;
}

export interface FoundryChatClient {
  completeStreamingChat(messages: Array<{ role: string; content: string }>): AsyncIterable<FoundryChunk>;
}

export function registerFoundryAgent(
  runtime: VifuRuntime,
  client: FoundryChatClient,
  options: { model: string; endpoint?: string },
): VifuRuntime {
  const endpoint = options.endpoint ?? "foundry-chat";
  return runtime.agent({
    id: "foundry-local",
    endpoint,
    metadata: { model: options.model, framework: "foundry-local" },
    handler: (request, trace) => invokeFoundry(request, trace, client, options.model),
  });
}

async function invokeFoundry(
  request: VifuAgentRequest,
  trace: VifuAgentTrace,
  client: FoundryChatClient,
  model: string,
) {
  const input = request.input as { prompt: string };
  const stream = client.completeStreamingChat([{ role: "user", content: input.prompt }]);
  const iterator = stream[Symbol.asyncIterator]();
  const parts: string[] = [];

  await trace.stage("first_token", async () => {
    while (true) {
      const next = await iterator.next();
      if (next.done) return;
      const content = next.value.choices?.[0]?.delta?.content;
      if (content) {
        parts.push(content);
        trace.outputDelta({ text: content });
        return;
      }
    }
  }, { model });

  await trace.stage("decode", async () => {
    while (true) {
      const next = await iterator.next();
      if (next.done) return;
      const content = next.value.choices?.[0]?.delta?.content;
      if (content) {
        parts.push(content);
        trace.outputDelta({ text: content });
      }
    }
  }, { model });

  return {
    output: { text: parts.join("") },
    metadata: { model, provider: "foundry-local" },
  };
}
